### PR TITLE
[WIP] implement connection broker

### PR DESCRIPTION
**Pull request by hkseo95**
Please review it. It is implemented by our first idea. The amqp usage was confusing, so I wanted to make it simple to use.

### CODE REVIEW

File: chamge/amqp-arbiter-backend.c [hunk 1]
```diff
@@ -38,59 +44,37 @@ chamge_amqp_arbiter_backend_enroll (ChamgeArbiterBackend * arbiter_backend)
...
-  amqp_queue_bind (self->amqp_conn, amqp_channel, amqp_r->queue,
-      amqp_cstring_bytes (amqp_exchange_name), queue, amqp_empty_table);
+  /* set amqp login information */
+  g_value_init (&value_str, G_TYPE_STRING);
+  g_value_init (&value_int, G_TYPE_UINT);
 
-  g_debug ("binding a queue (exchange: %s, bind_key: %s)", amqp_exchange_name,
-      amqp_enroll_q_name);
+  g_value_set_string (&value_str, amqp_uri);
+  g_object_set_property (G_OBJECT (self->amqp_rpc), "url", &value_str);
```
Comment (justinjoy) [inline chamge/amqp-arbiter-backend.c]: `g_object_set()`  doesn’t need `GValue`.

you might want to use `g_object_set (G_OBJECT (self->amqp_rpc), “url”, amqp_uri, NULL)`.

File: chamge/amqp-arbiter-backend.c [hunk 2]
```diff
@@ -107,18 +91,26 @@ _handle_edge_enroll (ChamgeAmqpArbiterBackend * self, JsonNode * root)
   JsonObject *json_object = NULL;
   const gchar *edge_id = NULL;
   ChamgeArbiterBackendClass *klass = CHAMGE_ARBITER_BACKEND_GET_CLASS (self);
+  gchar *body = NULL;;
```
Comment (justinjoy) [inline chamge/amqp-arbiter-backend.c]: `g_autofree gchar *body = NULL;`

Then, we can be freed from worrying about `g_free()`.

File: chamge/amqp-edge-backend.c [hunk 3]
```diff
@@ -7,19 +7,73 @@
 #include "config.h"
 
 #include "amqp-edge-backend.h"
+#include "amqp/amqp-rpc.h"
+
+#include <gio/gio.h>
+
+#define AMQP_EDGE_BACKEND_SCHEMA_ID "org.hwangsaeul.Chamge1.Arbiter.AMQP"
```
Comment (justinjoy) [inline chamge/amqp-edge-backend.c]: The id may be `org.hwangsaeul.Chamge1.Edge.AMQP`?

What I am thinking here is that *edge* shouldn’t know *arbiter* vice versa. 
they are solely independent.

File: chamge/amqp-edge-backend.c [hunk 4]
```diff
@@ -7,19 +7,73 @@
...
+#define AMQP_ARBITER_BACKEND_SCHEMA_ID "org.hwangsaeul.Chamge1.Arbiter.AMQP"
```
Comment (justinjoy) [inline chamge/amqp-edge-backend.c]: In that way, this shouldn’t be here.

File: chamge/amqp/amqp-rpc.c [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+{
+  gchar *correlation_id_str = NULL;
+  gchar *content_type_str = g_strdup (DEFAULT_CONTENT_TYPE);
+
+  props->_flags = AMQP_BASIC_CONTENT_TYPE_FLAG | AMQP_BASIC_DELIVERY_MODE_FLAG;
+  props->content_type = amqp_cstring_bytes (content_type_str);
+  props->delivery_mode = 2;     /* persistent delivery mode */
+
+  if (reply_queue_name != NULL) {
+    gchar *reply_to_queue = (gchar *) g_strdup (reply_queue_name);
```
Comment (justinjoy) [inline chamge/amqp/amqp-rpc.c]: no casting is required for `g_strdup()`.

File: chamge/amqp/amqp-rpc.c [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+  }
+}
+
+static void
+_amqp_parse_url (const gchar * url, rpc_connection_info * const conn_info)
+{
+  g_autofree gchar *s = g_strdup (url);
```
Comment (justinjoy) [inline chamge/amqp/amqp-rpc.c]: `s` isn’t used. it’s meaningless copy.

File: chamge/amqp/amqp-rpc.c [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+static void
+chamge_amqp_rpc_set_property (GObject * object,
+    guint prop_id, const GValue * value, GParamSpec * pspec)
+{
+  ChamgeAmqpRpc *self = CHAMGE_AMQP_RPC (object);
+  ChamgeAmqpRpcPrivate *priv = chamge_amqp_rpc_get_instance_private (self);
+
+  switch ((_ChamgeAmqpRpcProperty) prop_id) {
+    case PROP_HOSTNAME:
+      priv->connection_info.host = g_value_dup_string (value);
```
Comment (justinjoy) [inline chamge/amqp/amqp-rpc.c]: this way will cause memory leakage if user sets more than once. 
`g_free()` should be called before copying string.

File: chamge/amqp/amqp-rpc.c [hunk 8]
```diff
@@ -0,0 +1,700 @@
...
+
+static void
+chamge_amqp_rpc_class_init (ChamgeAmqpRpcClass * klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->get_property = chamge_amqp_rpc_get_property;
+  object_class->set_property = chamge_amqp_rpc_set_property;
+  object_class->dispose = chamge_amqp_rpc_dispose;
+  object_class->finalize = chamge_amqp_rpc_finalize;
```
Comment (justinjoy) [inline chamge/amqp/amqp-rpc.c]: if there’s no specific reason, let’s override `_dispose()` only. 
Normally, it’s enough to free resources in dispose.

File: chamge/amqp/amqp-rpc.c [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+  properties[PROP_VHOST] = g_param_spec_string ("vhost", "vhost", "vhost name",
+      NULL, G_PARAM_READWRITE);
+
+  properties[PROP_ID] =
+      g_param_spec_string ("id", "id", "use id for login amqp broker", NULL,
+      G_PARAM_READWRITE);
+
+  properties[PROP_PWD] =
+      g_param_spec_string ("pwd", "pwd", "password for login amqp broker", NULL,
+      G_PARAM_READWRITE);
```
Comment (justinjoy) [inline chamge/amqp/amqp-rpc.c]: password shouldn’t be readable.

File: chamge/amqp/amqp-rpc.h [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+#ifndef __CHAMGE_AMQP_RPC_H__
+#define __CHAMGE_AMQP_RPC_H__
+
+#include <gio/gio.h>
+#include <chamge/types.h>
+
+G_BEGIN_DECLS
+#define CHAMGE_TYPE_AMQP_RPC       (chamge_amqp_rpc_get_type ())
+    CHAMGE_API_EXPORT
```
Comment (justinjoy) [inline chamge/amqp/amqp-rpc.h]: I guess this object is only for internal use. then, we don’t need to export these apis.

File: chamge/amqp/amqp-rpc.h [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+G_DECLARE_DERIVABLE_TYPE (ChamgeAmqpRpc, chamge_amqp_rpc, CHAMGE, AMQP_RPC,
```
Comment (justinjoy) [inline chamge/amqp/amqp-rpc.h]: This class doesn’t have virtual method. that means it would’t be derivable.

File: chamge/amqp/amqp-rpc.h [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    GObject)
+     struct _ChamgeAmqpRpcClass
+     {
+       GObjectClass parent_class;
+
+     };
```
Comment (justinjoy) [inline chamge/amqp/amqp-rpc.h]: do not run indent for header.

File: chamge/amqp-arbiter-backend.c [hunk 13]
```diff
@@ -7,21 +7,27 @@
...
+} AmqpReturnAuguments;
 
 struct _ChamgeAmqpArbiterBackend
 {
   ChamgeArbiterBackend parent;
   GSettings *settings;
 
-  amqp_connection_state_t amqp_conn;
-  amqp_socket_t *amqp_socket;
+  AmqpReturnAuguments amqp_return_arg;
```
Comment (justinjoy) [inline chamge/amqp-arbiter-backend.c]: `reply_queue` and `correlation_id` can be members here. 
but not sure if they need to be shared.
 argument passing is more maintainable than holding and sharing pointers.

File: chamge/amqp-arbiter-backend.c (hunk 1 above)
Comment (hkseo95) [reply chamge/amqp-arbiter-backend.c]: Yes. Thanks!!

File: chamge/amqp-arbiter-backend.c (hunk 2 above)
Comment (hkseo95) [reply chamge/amqp-arbiter-backend.c]: Yes.

File: chamge/amqp-edge-backend.c (hunk 3 above)
Comment (hkseo95) [reply chamge/amqp-edge-backend.c]: I know it. I was just lazy. ^^; I will fix it, too.

File: chamge/amqp-edge-backend.c (hunk 4 above)
Comment (hkseo95) [reply chamge/amqp-edge-backend.c]: Yes!

File: chamge/amqp/amqp-rpc.c (hunk 5 above)
Comment (hkseo95) [reply chamge/amqp/amqp-rpc.c]: Oops! ^^;

File: chamge/amqp/amqp-rpc.c (hunk 9 above)
Comment (hkseo95) [reply chamge/amqp/amqp-rpc.c]: Yes. I will fix it.

File: chamge/amqp/amqp-rpc.h (hunk 10 above)
Comment (hkseo95) [reply chamge/amqp/amqp-rpc.h]: Yes.

File: chamge/amqp-arbiter-backend.c (hunk 13 above)
Comment (hkseo95) [reply chamge/amqp-arbiter-backend.c]: Yes. You are right. I will fix it, too.

**Pull request closed without merging**